src/query/sql/src/planner/optimizer/optimizer.rs

```diff
@@ -127,14 +127,28 @@ pub fn optimize(
             Ok(Plan::CopyIntoTable(plan))
         }
         Plan::MergeInto(plan) => {
+            // optimize source :fix issue #13733
+            let right_source = optimize_query(
+                ctx.clone(),
+                opt_ctx.clone(),
+                plan.meta_data.clone(),
+                plan.input.child(1)?.clone(),
+            )?;
+            // replace right source
```

Comment:
Why are we replacing the right source? better to add some comments to explain the reason.

Reply:
for example the sql in https://github.com/datafuselabs/databend/issues/13733, there is subquery expr like `CASE 
            WHEN number % 2 = 0 THEN (number / 2) % 250000
            ELSE (SELECT MAX(order_id) FROM orders) + number + 1
        END AS order_id,`, so we will get invalid index in https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/semantic/lowering.rs#L234 ,the same with https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/semantic/lowering.rs#L205,https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/semantic/lowering.rs#L180,etc. So we will use optimizer to solve the index